amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpAsyncFlow.scala

```diff
@@ -0,0 +1,330 @@
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        upstreamException = Some(ex)
+        if (isFinished)
+          closeStage()
+        else
+          log.debug("Received upstream failure signal - stage will be failed when all buffered messages are processed")
+
+      }
+
+      override def onUpstreamFinish(): Unit =
+        if (noAwaitingMessages && exitQueue.isEmpty) {
+          promise.success(Done)
+          super.onUpstreamFinish()
+        } else {
+          log.debug("Received upstream finish signal - stage will be closed when all buffered messages are processed")
+        }
+
+      private def publish(message: WriteMessage[T]): DeliveryTag = {
+        val tag: DeliveryTag = channel.getNextPublishSeqNo
+
+        log.debug("Publishing message {} with deliveryTag {}.", message, tag)
+
+        channel.basicPublish(
+          exchange,
+          message.routingKey.getOrElse(routingKey),
+          message.mandatory,
+          message.immediate,
+          message.properties.orNull,
+          message.bytes.toArray
+        )
+
+        tag
+      }
+    }
+  )
+
+  setHandler(
+    out,
+    new OutHandler {
+      override def onPull(): Unit = {
+        if (exitQueue.nonEmpty) {
+          val result = exitQueue.dequeue()
+          log.debug("Pushing enqueued {} downstream.", result)
+          push(out, result)
+        }
+
+        if (isFinished) closeStage()
+        else if (!hasBeenPulled(in)) tryPull(in)
+      }
+    }
+  )
+
+  override protected def onTimer(timerKey: Any): Unit =
+    timerKey match {
+      case tag: DeliveryTag => {
+        log.debug("Received timeout for deliveryTag {}.", tag)
+        onRejection(tag, multiple = false)
+      }
+      case _ => ()
+    }
+
+  private def closeStage(): Unit =
+    upstreamException match {
+      case Some(throwable) =>
+        promise.failure(throwable)
+        failStage(throwable)
+      case None =>
+        promise.success(Done)
+        completeStage()
+    }
+
+  private def isFinished: Boolean = isClosed(in) && noAwaitingMessages && exitQueue.isEmpty
+}
+
+/**
+ * Internal API.
+ *
+ * AMQP flow that uses asynchronous confirmations but preserves the order in which messages were pulled.
+ * Internally messages awaiting confirmation are stored in a ordered buffer. Initially messages have `ready`
+ * flag set to `false`. On confirmation flag is changed to `true` and messages from the beginning of the queue
+ * are dequeued in sequence until first non-ready message - this means that if confirmed message is not a
+ * first element of the buffer then nothing will be dequeued. Flag `multiple` on a confirmation means that
+ * broker confirms all messages up to a given delivery tag, which means that so all messages up to (and including)
+ * this delivery tag can be safely dequeued.
+ */
+@InternalApi private[amqp] final class AmqpAsyncFlow[T](
```

Comment:
Please append `Stage` to the stage class names. We can do without `sealed`.

Reply:
I changed it, although I'm not sure now what is the convenion. In reference connector there are no `Stage` suffixes 😛 .